src/lib.rs

```diff
@@ -1,8 +1,278 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        let result = 2 + 2;
-        assert_eq!(result, 4);
+#![no_std]
+/// MQTT Request/response Handling
+///
+/// # Overview
+/// This library is intended to be an easy way to handle inbound requests automatically.
+///
+/// Handler functions can be associated with the library to be automatically called whenever a
+/// specified request is received, and the handler will automatically be invoked when the request
+/// is received.
+///
+/// ## Limitations
+/// * The `poll()` function has a somewhat odd signature (using a function to provide the `Context`
+/// and call the handler) due to required compatibility with RTIC and unlocked resources.
+///
+/// * Handlers may only be closures that do not capture any local resources. Instead, move local
+/// captures into the `Context`, which will be provided to the handler in the function call.
+///
+use core::fmt::Write;
+
+use minimq::{
+    embedded_nal::{IpAddr, TcpClientStack},
+    embedded_time, Property, QoS, Retain,
+};
+
+use serde_json_core::heapless::{String, Vec};
+
+use log::{info, warn};
+
+pub mod response;
+pub use response::Response;
+
+// The maximum topic length of any settings path.
+const MAX_TOPIC_LENGTH: usize = 128;
+
+// The keepalive interval to use for MQTT in seconds.
+const KEEPALIVE_INTERVAL_SECONDS: u16 = 60;
+
+#[derive(Debug, PartialEq)]
+pub enum Error<E> {
+    RegisterFailed,
+    PrefixTooLong,
+    Deserialization(serde_json_core::de::Error),
+    Mqtt(minimq::Error<E>),
+}
+
+impl<E> From<serde_json_core::de::Error> for Error<E> {
+    fn from(e: serde_json_core::de::Error) -> Self {
+        Error::Deserialization(e)
+    }
+}
+
+impl<E> From<minimq::Error<E>> for Error<E> {
+    fn from(e: minimq::Error<E>) -> Self {
+        Error::Mqtt(e)
+    }
+}
+
+mod sm {
+    smlang::statemachine! {
+        transitions: {
+            *Init + Connected = Connected,
+            Connected + Subscribed = Active,
+            _ + Reset = Init,
+        }
+    }
+
+    pub struct Context;
+
+    impl StateMachineContext for Context {}
+}
+
+type Handler<Context, E> = fn(&mut Context, &str, &[u8]) -> Result<Response, Error<E>>;
+
+/// MQTT request/response interface.
+pub struct Minireq<Context, Stack, Clock, const MESSAGE_SIZE: usize>
+where
+    Stack: TcpClientStack,
+    Clock: embedded_time::Clock,
+{
+    // TODO: Maybe use a hash-based IndexMap?
+    requests: heapless::LinearMap<String<60>, Handler<Context, Stack::Error>, 10>,
```

Comment:
I've updated the number of requests (10) to be generic. I'm not really keen on also making the maximum command length generic because of type bloat (at least until default const-generics stabilizes), so I'd prefer we just use some fixed max size. I've upped it to 128. This size does _not_ include the prefix len, so that should be plenty of size for any use case I can think of.

Reply:
Ack. But then 60 or 64 seems adequate for command length. This is only the command topic, not the payload.